iceoryx_utils/test/moduletests/test_cxx_unique_ptr.cpp

```diff
@@ -0,0 +1,433 @@
+{
+  public:
+    void SetUp()
+    {
+        m_deleterCalled = false;
+        m_anotherDeleterCalled = false;
+    }
+
+    void TearDown()
+    {
+    }
+
+    bool m_deleterCalled;
+    bool m_anotherDeleterCalled;
+
+    std::function<void(Position* const)> deleter = [this](Position* const p) {
+        m_deleterCalled = true;
+        delete p;
+    };
+
+    std::function<void(Position* const)> anotherDeleter = [this](Position* const p) {
+        m_anotherDeleterCalled = true;
+        delete p;
+    };
+};
+
+TEST_F(UniquePtrTest, CtorWithNullptrSetsPtrToNullAndDoesntCallDeleter)
+{
+    {
+        auto sut = iox::cxx::unique_ptr<Position>(nullptr);
+        EXPECT_FALSE(sut);
+        EXPECT_EQ(sut.get(), nullptr);
+    }
+    // SUT is out of scope but shouldn't have called deleter as SUT is NULL
+    EXPECT_FALSE(m_deleterCalled);
+}
+
+TEST_F(UniquePtrTest, CtorWithOnlyDeleterSetsPtrToNullAndDoesntCallDeleter)
+{
+    {
+        auto sut = iox::cxx::unique_ptr<Position>(deleter);
+        EXPECT_FALSE(sut);
+        EXPECT_EQ(sut.get(), nullptr);
+    }
+    // SUT is out of scope but shouldn't have called deleter as SUT is NULL
+    EXPECT_FALSE(m_deleterCalled);
+}
+
+TEST_F(UniquePtrTest, CtorWithObjectPtrAndDeleterSetsPtrToObjectAndCallsDeleter)
+{
+    {
+        auto object = new Position();
+        auto sut = iox::cxx::unique_ptr<Position>(object, deleter);
+        EXPECT_TRUE(sut);
+        EXPECT_EQ(sut.get(), object);
+    }
+    // SUT is out of scope and should have called deleter
+    EXPECT_TRUE(m_deleterCalled);
+}
+
+TEST_F(UniquePtrTest, CtorWithObjectPtrToNullAndDeleterSetsPtrToObjectAndDoesntCallsDeleter)
+{
+    {
+        auto sut = iox::cxx::unique_ptr<Position>(nullptr, deleter);
+        EXPECT_FALSE(sut);
+        EXPECT_EQ(sut.get(), nullptr);
+    }
+    // SUT is out of scope but shouldn't have called deleter as SUT is NULL
+    EXPECT_FALSE(m_deleterCalled);
+}
+
+TEST_F(UniquePtrTest, CtorUsingMoveWithObjectPtrAndDeleterSetsPtrToObjectAndCallsDeleter)
+{
+    {
+        auto object = new Position();
+        auto sut = iox::cxx::unique_ptr<Position>(object, deleter);
+
+        auto anotherSut = iox::cxx::unique_ptr<Position>(std::move(sut));
+
+        // no deleter call during move
+        EXPECT_FALSE(m_deleterCalled);
+        EXPECT_FALSE(sut);
+        EXPECT_EQ(anotherSut.get(), object);
+    }
+    // SUT is out of scope and should have called deleter
+    EXPECT_TRUE(m_deleterCalled);
+}
+
+TEST_F(UniquePtrTest, MoveAssignmentUniquePtrsSetsPtrToObjectAndCallsDeleter)
+{
+    {
+        auto object = new Position();
+        auto sut = iox::cxx::unique_ptr<Position>(object, deleter);
+
+        auto anotherSut = std::move(sut);
+
+        // no deleter call during movie
+        EXPECT_FALSE(m_deleterCalled);
+        EXPECT_FALSE(sut);
+        EXPECT_EQ(anotherSut.get(), object);
+    }
+    // anotherSUT is out of scope and should have called deleter
+    EXPECT_TRUE(m_deleterCalled);
```

Comment:
You could verify that the deleter wasn't called twice?

Reply:
Done